File: test/integration/tests/l7/cgi-bin/redirect_login.lua
-- integration/tests/l7/cgi-bin/redirect_login.lua
-- Simulates a login CGI that returns 302 redirect with session cookie

-- Read Environment Variables
local method = os.getenv("REQUEST_METHOD") or "GET"
local content_length = os.getenv("CONTENT_LENGTH") or "0"

-- Read POST Body (simulated form data)
local body = ""
local len = tonumber(content_length)
if len and len > 0 then
	body = io.read(len) or ""
end

-- Debug: Log what we received (to stderr, visible in Vane debug logs)
io.stderr:write(string.format("DEBUG: method=%s, content_length=%s, body_len=%d\n",
	method, content_length, #body))

-- Simulate Authentication (check if body contains credentials)
local authenticated = false
if method == "POST" and #body > 0 then
	-- Check if body contains username and password fields
	if body:find("username=") and body:find("password=") then
		authenticated = true
		io.stderr:write("DEBUG: Authentication SUCCESS\n")
	else
		io.stderr:write("DEBUG: Authentication FAILED - invalid body format\n")
	end
else
	io.stderr:write(string.format("DEBUG: Authentication FAILED - method=%s, body_len=%d\n", method, #body))
end

if authenticated then
	-- Output 302 Redirect with Session Cookie
	print("Status: 302 Found")
	print("Set-Cookie: session_id=test_session_12345; path=/; HttpOnly")
	print("Location: /dashboard")
	print("Content-Type: text/html")
	print("") -- End of Headers
	-- Optional: 302 can have empty body or minimal body
	print("<html><body>Redirecting...</body></html>")
else
	-- Return 401 Unauthorized for missing/invalid credentials
	print("Status: 401 Unauthorized")
	print("Content-Type: text/plain")
	print("") -- End of Headers
	print("Authentication required")
end

File: test/integration/tests/l7/cgi-bin/redirect_login.c
/* integration/tests/l7/cgi-bin/redirect_login.c */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    // 1. Read Environment Variables
    char *method = getenv("REQUEST_METHOD");
    char *content_length_str = getenv("CONTENT_LENGTH");

    // Debug to stderr (visible in Vane logs)
    fprintf(stderr, "DEBUG: method=%s, content_length=%s\n",
            method ? method : "(null)",
            content_length_str ? content_length_str : "(null)");

    // 2. Read POST Body
    char body[4096] = {0};
    int len = 0;
    if (content_length_str != NULL) {
        len = atoi(content_length_str);
        if (len > 0) {
            if (len > sizeof(body) - 1) len = sizeof(body) - 1;
            fread(body, 1, len, stdin);
            fprintf(stderr, "DEBUG: Read %d bytes from stdin\n", len);
        }
    }

    // 3. Check Authentication (simple string match)
    int authenticated = 0;
    if (method && strcmp(method, "POST") == 0 && len > 0) {
        // Check if body contains "username=" and "password="
        if (strstr(body, "username=") != NULL && strstr(body, "password=") != NULL) {
            authenticated = 1;
            fprintf(stderr, "DEBUG: Authentication SUCCESS\n");
        } else {
            fprintf(stderr, "DEBUG: Authentication FAILED - invalid credentials\n");
        }
    } else {
        fprintf(stderr, "DEBUG: Authentication FAILED - not POST or no body\n");
    }

    // 4. Output Response
    if (authenticated) {
        // Return 302 Redirect with Session Cookie
        printf("Status: 302 Found\r\n");
        printf("Set-Cookie: session_id=test_session_12345; path=/; HttpOnly\r\n");
        printf("Location: /dashboard\r\n");
        printf("Content-Type: text/html\r\n");
        printf("\r\n");
        // Optional body for 302
        printf("<html><body>Redirecting...</body></html>\n");
    } else {
        // Return 401 Unauthorized
        printf("Status: 401 Unauthorized\r\n");
        printf("Content-Type: text/plain\r\n");
        printf("\r\n");
        printf("Authentication required\n");
    }

    fflush(stdout);
    return 0;
}
